analysis/type.go

```diff
@@ -73,7 +73,89 @@ type TokenFilter interface {
 }
 
 type Analyzer interface {
-	Analyze([]byte) TokenStream
+	Type() string
+	// return value of this method depends on the type of analyzer
+	Analyze([]byte) interface{}
+}
+
+const (
+	TokensAnalyzerType     = "token"
+	HookTokensAnalyzerType = "hook_token"
+	VectorAnalyzerType     = "vector"
+	HookVectorAnalyzerType = "hook_vector"
+)
+
+// # Analyzer Analyze() return type
+type TokensAnalyzer struct {
+	Tokens TokenStream
+}
+type HookTokensAnalyzer struct {
+	Tokens TokenStream
+	Err    error
+}
+type VectorAnalyzer []float64
+type HookVectorAnalyzer struct {
+	Vector []float64
+	Err    error
+}
+
+func AnalyzeForTokens(analyzer Analyzer, input []byte) (TokenStream, error) {
+	analyzerType := analyzer.Type()
+	if analyzerType != TokensAnalyzerType &&
+		analyzerType != HookTokensAnalyzerType {
+		return nil, fmt.Errorf("cannot analyze text with analyzer of type: %s",
```

Comment:
alternate error msg: "given analyzer is not compatible to be used as a token analyzer"